Add support for Unix sockets

fpagliughi opened the pull request:
Adds support for UNIX-domain sockets on *nix platforms.
It implements issue #864.

This allows clients to communicate with an MQTT server over UNIX-domain sockets on *nix systems. (Not on Windows). This is opt-in with a CMake build option, `PAHO_WITH_UNIX_SOCKETS`.

This can increase performance and reduce latency since UNIX sockets have a much simpler network stack than TCP. But perhaps more importantly, it provides increased security with more fine-grained control when clients reside on the same host as the broker. Access to the broker can be done using file permissions (with groups, owners, etc).

The clients can request a socket of this type using a URI of `unix:///path/to/socket/file`. The socket file is created by the server when it binds and listens on the UNIX socket.

The implementation here is fairly trivial. The library must now recognize the `unix://` scheme in the URI, then create and connect to a socket using a *sockaddr_un* socket address containing the path to the socket file.

After that all communications and closing the socket are handled by the same code that manages TCP connections. A socket is a socket.

This was tested on Linux and macOS with the _mosquitto_ server. By definition it requires a broker running on the same host whereby it communicates with the clients using the local file system.

The mosquitto configuration file can be as simple as:
```
allow_anonymous true    
listener 0 /tmp/mosquitto.sock
```

This PR also updates several of the simple sample apps to take a URI at the command line, allowing them to be used to try this out, like:
```
MQTTAsync_publish unix:///tmp/mosquitto.sock
```

vjardin commented on src/MQTTProtocolOut.c:
```diff
@@ -211,18 +211,18 @@ int MQTTProtocol_setHTTPProxy(Clients* aClient, char* source, char** dest, char*
 #if defined(OPENSSL)
 #if defined(__GNUC__) && defined(__linux__)
-int MQTTProtocol_connect(const char* ip_address, Clients* aClient, int ssl, int websocket, int MQTTVersion,
+int MQTTProtocol_connect(const char* ip_address, Clients* aClient, int unixsock, int ssl, int websocket, int MQTTVersion,
```
`ip_address` is not just an IP address, it can be a path to a unix socket, please, can you rename it accordingly ?

vjardin commented on src/MQTTAsync.c:
```diff
@@ -329,6 +329,9 @@ int MQTTAsync_createWithOptions(MQTTAsync* handle, const char* serverURI, const
 	{
 		if (strncmp(URI_TCP, serverURI, strlen(URI_TCP)) != 0
 		 && strncmp(URI_MQTT, serverURI, strlen(URI_MQTT)) != 0
+#if defined(UNIXSOCK)
```
you should avoid such `if defined()` in order to avoid code complexity. The error cases can be handled for WIN32 or non UNIXSOCK into the function `Socket_unix_new()` that can be available for all the cases, but that should return a runtime ERROR when Unixsock is not available. 

So the code will be easier to read.

fpagliughi replied:
I disagree here. The conditional compilation ensures that `unix://` is reported as an unsupported protocol if support is not compiled into the library - just the same way it's done with  `ssl://`, `mqtts://`, and `wss://`.

vjardin commented on src/MQTTAsync.c:
```diff
@@ -384,6 +387,13 @@ int MQTTAsync_createWithOptions(MQTTAsync* handle, const char* serverURI, const
 		serverURI += strlen(URI_TCP);
 	else if (strncmp(URI_MQTT, serverURI, strlen(URI_MQTT)) == 0)
 		serverURI += strlen(URI_MQTT);
+#if defined(UNIXSOCK)
```
same, remove `#if defined(UNIXSOCK)`

vjardin commented on src/Socket.c:
```diff
@@ -1356,6 +1361,58 @@ int Socket_new(const char* addr, size_t addr_len, int port, SOCKET* sock)
 	return rc;
 }
 
+#if defined(UNIXSOCK)
```
No, this `int Socket_unix_new(const char* addr, size_t addr_len, SOCKET* sock)` should be available for all OS ; but in case of WIN32, it should return a PAHO error or a SOCKET_ERROR with a log that Unix socket is not available.

Doing so, you'll remove many `#if defined(UNIXSOCK).

fpagliughi replied:
I agree with you that all the conditional compilation in these files makes it quite difficult to follow the flow through this code. But I think, at this point, it is probably be better to be consistent with existing conventions than start doing things differently for a small portion of the code. And, realistically, the inline code for this feature is trivial compared to the rest of what's there.

vjardin commented on src/samples/MQTTClient_publish.c:
```diff
@@ -34,7 +34,10 @@ int main(int argc, char* argv[])
     int rc;
 
-    if ((rc = MQTTClient_create(&client, ADDRESS, CLIENTID,
+	const char* uri = (argc > 1) ? argv[1] : ADDRESS;
```
ident issue.

vjardin commented on src/samples/MQTTClient_publish_async.c:
```diff
@@ -65,7 +65,10 @@ int main(int argc, char* argv[])
     int rc;
 
-    if ((rc = MQTTClient_create(&client, ADDRESS, CLIENTID,
+	const char* uri = (argc > 1) ? argv[1] : ADDRESS;
```
ident issue.

fpagliughi replied:
Argh. Yeah, you're correct. There is inconsistent formatting in this project. Why are the samples using spaces when the rest of the code is using tabs?

vjardin commented on src/samples/MQTTAsync_subscribe.c:
```diff
@@ -137,7 +137,10 @@ int main(int argc, char* argv[])
 	int ch;
 
-	if ((rc = MQTTAsync_create(&client, ADDRESS, CLIENTID, MQTTCLIENT_PERSISTENCE_NONE, NULL))
+	const char* uri = (argc > 1) ? argv[1] : ADDRESS;
```
ident issue

vjardin commented on src/samples/MQTTAsync_publish.c:
```diff
@@ -146,7 +146,10 @@ int main(int argc, char* argv[])
 	int rc;
 
-	if ((rc = MQTTAsync_create(&client, ADDRESS, CLIENTID, MQTTCLIENT_PERSISTENCE_NONE, NULL)) != MQTTASYNC_SUCCESS)
+	const char* uri = (argc > 1) ? argv[1] : ADDRESS;
```
ident issue

vjardin commented on src/Socket.c:
```diff
@@ -1276,14 +1281,14 @@ int Socket_new(const char* addr, size_t addr_len, int port, SOCKET* sock)
-						Log(LOG_ERROR, -1, "Could not set SO_SNDBUF for socket %d", *sock);
-				}
-	#endif
+			if (1)
+			{
+				int optsend = 100; //2 * 1440;
+				printf("Setting optsend to %d\n", optsend);
+				if (setsockopt(*sock, SOL_SOCKET, SO_SNDBUF, (void*)&optsend, sizeof(optsend)) != 0)
+					Log(LOG_ERROR, -1, "Could not set SO_SNDBUF for socket %d", *sock);
+			}
+#endif
```
you are not changing any code -> ident issue. The diff should be empty.

fpagliughi replied:
I disagree in the sense that the previous indenting was incorrect and misleading. Admittedly, this has nothing to do with the rest of the PR; it's a holdover from my initial attempt to modify `Socket_new()` for UNIX sockets, before deciding that adding a separate `Socket_unix_new()` would be more clear.

I'd prefer to keep it.

vjardin reviewed (changes requested):
Thansk @fpagliughi ; it'll be a nice security and performance improvement. I took the liberty to review your commits.

fpagliughi commented on src/MQTTAsyncUtils.h:
```diff
@@ -25,6 +25,10 @@
 #define URI_WS   "ws://"
 #define URI_WSS  "wss://"
 
+#if defined(UNIXSOCK)
```
True enough.

fpagliughi commented on src/MQTTProtocolOut.c:
```diff
@@ -318,6 +318,12 @@ int MQTTProtocol_connect(const char* ip_address, Clients* aClient, int websocket
+#if defined(UNIXSOCK)
+	else if (unixsock) {
+		addr_len = strlen(ip_address);
+		rc = Socket_unix_new(ip_address, addr_len, &(aClient->net.socket));
```
I was about to argue against doing this (not change too much), but the change is isolated and makes sense. Also, then, the function comments need to be updated.

fpagliughi commented:
@vjardin Thanks so much for the review! So nice to have this quick feedback. I made some changes, and added my reasoning for not making other changes.

vjardin commented:
OK, LGTM.

Regaarding NOSIGPIPE, if it is not defined, it means that you should handle the SIGPIPE, I did not check if it is properly done. But in 2025, it should not be a topic !

fpagliughi commented:
For years I made the argument to implement UNIX sockets for MQTT to reduce latency in local systems, but that never seemed to sway people. The security argument gets people's attention.

But I just ran my first test, compiling this branch into the Paho Rust library and using it on an MQTT v5 RPC system that I have for one of my customers. Each service exports a _ping_ function to check that it's alive and responsive. The full round-trip is four publishes:
```
client [request] -> broker -> service [response] -> broker -> client
```
I just tried it to compare _localhost_ TCP to Unix sockets:
```
$ svcctl ping  -U tcp://localhost:1883                                   
Service responded in 43.247ms

$ svcctl ping -U unix:///tmp/mosquitto.sock
Service responded in 2.530ms
```

Using TCP takes 43ms... using UNIX sockets takes 2.5ms.

That's a 94% reduction in round-trip latency!!!

@icraggs :  I need this mergerd and released ASAP!!!    :-)

vjardin commented:
Same, this requirement of unix socket is for both security and performances.

icraggs commented:
Great! Does this change the external interfaces - as in could we have it in the next 1.3.n release?

I'm close to getting a 1.3.14 release out (I'm doing a proxy issue at the moment ) @fpagliughi could you take a look at this build issue to help me out? https://github.com/eclipse-paho/paho.mqtt.c/issues/1376

fpagliughi commented:
Hey @icraggs, no this doesn't change the external interfaces. When built in, it just allows the library to recognize a new URI scheme ("unix://") and handle it accordingly. I would love to get it into v1.3.x

I will squash the commits in this PR to make it easy to apply to both branches, and run some tests myself to merge and build in the `develop` branch.

fpagliughi commented:
Later today I can get a Windows machine running and test #1376

fpagliughi commented:
OK, I squashed this PR and then tried to cherry-pick the resulting commit into the `develop` branch. It didn't merge cleanly due to the choose-an-interface code from v1.4 in _Socket.c_ 

So I cleaned up the merge and created another pull request, #1553.

So **both** PRs can be landed into their target branches... this one into v1.4 and the other into develop.

icraggs commented:
I use the paho_c_pub and paho_c_sub samples a lot. Their interface is similar to mosquitto_pub and sub. You can try the unix sockets without any changes, for instance:

`build.paho/src/samples/paho_c_pub x --connection unix:///tmp/mosquitto.sock --trace protocol`
